File: jadelens/unified_diff.py
"""Parsing and application of 0-context unified diffs.

This is a custom implementation because we want strict, predictable
semantics rather than the fuzzy / line-number-drift behaviour of GNU
``patch`` and most off-the-shelf libraries:

- All hunks within a diff reference the **original** file's line numbers.
  The bot reads the file once, plans multiple edits against that snapshot,
  and emits one diff with multiple hunks.
- Hunks must be in strictly increasing ``-L`` order and must not overlap.
- Each ``-`` line in a hunk must match the file's line at its claimed
  position exactly, byte-for-byte (modulo the trailing newline). Any
  mismatch aborts the entire diff with an informative error.
- Lines are assumed to be \\n-terminated. The applier preserves the
  original file's trailing-newline status; added lines get \\n appended.
"""

import re
from dataclasses import dataclass

HUNK_HEADER_RE = re.compile(
    r"^@@ -(\d+)(?:,(\d+))? \+\d+(?:,(\d+))? @@"
)


class DiffParseError(Exception):
    """The diff text could not be parsed as a 0-context unified diff."""


class DiffApplyError(Exception):
    """The diff parsed but couldn't be applied (verification mismatch, etc.)."""


@dataclass(slots=True, frozen=True)
class Hunk:
    old_start: int  # 1-indexed line in old file; for pure insert, the line AFTER which to insert
    old_count: int  # number of lines removed (0 for pure insert)
    removed_lines: tuple[str, ...]
    added_lines: tuple[str, ...]


def parse_unified_diff(diff_text: str) -> list[Hunk]:
    """Parse a 0-context unified diff into a list of Hunks.

    Tolerates the optional ``--- a/...`` / ``+++ b/...`` header lines (and
    any other lines before the first ``@@``). Raises ``DiffParseError`` on
    format issues.
    """
    lines = diff_text.split("\n")
    # If the diff ended with \n, split gives a trailing empty string. Drop it.
    if lines and lines[-1] == "":
        lines.pop()

    hunks: list[Hunk] = []
    i = 0
    # Skip any preamble (file headers, etc.) until the first hunk header.
    while i < len(lines) and not lines[i].startswith("@@"):
        i += 1

    while i < len(lines):
        header_match = HUNK_HEADER_RE.match(lines[i])
        if not header_match:
            raise DiffParseError(f"Malformed hunk header: {lines[i]!r}")
        old_start = int(header_match.group(1))
        old_count = (
            int(header_match.group(2)) if header_match.group(2) is not None else 1
        )
        new_count = (
            int(header_match.group(3)) if header_match.group(3) is not None else 1
        )
        i += 1

        removed: list[str] = []
        added: list[str] = []
        while i < len(lines) and not lines[i].startswith("@@"):
            line = lines[i]
            if line.startswith("-"):
                removed.append(line[1:])
            elif line.startswith("+"):
                added.append(line[1:])
            elif line.startswith(" "):
                raise DiffParseError(
                    f"Context lines are not supported (0-context format "
                    f"required): {line!r}"
                )
            else:
                raise DiffParseError(
                    f"Unexpected line in hunk body: {line!r}"
                )
            i += 1

        if len(removed) != old_count:
            raise DiffParseError(
                f"Hunk header declared {old_count} removed line(s) but "
                f"hunk body has {len(removed)}"
            )
        if len(added) != new_count:
            raise DiffParseError(
                f"Hunk header declared {new_count} added line(s) but "
                f"hunk body has {len(added)}"
            )

        hunks.append(
            Hunk(
                old_start=old_start,
                old_count=old_count,
                removed_lines=tuple(removed),
                added_lines=tuple(added),
            )
        )

    _verify_ordering_and_non_overlap(hunks)
    return hunks


def _verify_ordering_and_non_overlap(hunks: list[Hunk]) -> None:
    """Hunks must be strictly increasing in old_start AND not touch positions a
    previous hunk already addressed.

    The minimum allowed ``B.old_start`` depends on both A's type (addressing
    vs. pure insert) and B's type, because pure inserts occupy a point
    *between* lines:

    | A           | B           | min B.old_start              |
    |-------------|-------------|------------------------------|
    | addressing  | addressing  | A.old_start + A.old_count    |
    | addressing  | pure insert | A.old_start + A.old_count - 1|
    | pure insert | addressing  | A.old_start + 1              |
    | pure insert | pure insert | A.old_start + 1              |

    The addressing-then-pure-insert case is the subtle one: an insert at the
    boundary just after A's last addressed line is legitimate (e.g. replace
    line 5, then insert after line 5).
    """
    for i in range(len(hunks) - 1):
        a, b = hunks[i], hunks[i + 1]
        if a.old_count == 0:
            # A is a pure insert; B (any type) must start strictly after.
            min_b_start = a.old_start + 1
        elif b.old_count == 0:
            # A addressing, B pure insert: B at the boundary right after
            # A's last addressed line is OK.
            min_b_start = a.old_start + a.old_count - 1
        else:
            # Both addressing: B must start after A's range.
            min_b_start = a.old_start + a.old_count
        if b.old_start < min_b_start:
            raise DiffParseError(
                f"Hunks out of order or overlapping: previous hunk "
                f"@@ -{a.old_start},{a.old_count} ... @@ is followed by "
                f"hunk @@ -{b.old_start},{b.old_count} ... @@ "
                f"(B.old_start must be >= {min_b_start})"
            )


def apply_unified_diff(original: str, diff_text: str) -> str:
    """Apply a 0-context unified diff to ``original``, returning the new text.

    Raises ``DiffParseError`` if the diff is malformed, or ``DiffApplyError``
    if any hunk's removed lines don't match the file at the claimed position.
    """
    hunks = parse_unified_diff(diff_text)
    original_lines = original.splitlines(keepends=True)

    new_lines: list[str] = []
    cursor = 0  # 0-indexed position in original_lines

    for hunk in hunks:
        # Convert 1-indexed old_start to 0-indexed position in original_lines.
        if hunk.old_count > 0:
            insert_at = hunk.old_start - 1
        else:
            # Pure insert: hunk header's old_start is the line AFTER which
            # to insert. For old_start=0, insert at index 0 (start of file).
            insert_at = hunk.old_start

        # Copy untouched lines from cursor up to the hunk's position.
        new_lines.extend(original_lines[cursor:insert_at])

        # Verify removed lines match file content at the claimed position.
        for offset, expected in enumerate(hunk.removed_lines):
            idx = insert_at + offset
            if idx >= len(original_lines):
                raise DiffApplyError(
                    f"Hunk references line {idx + 1} but file only has "
                    f"{len(original_lines)} line(s)"
                )
            actual = original_lines[idx].rstrip("\n")
            if actual != expected:
                raise DiffApplyError(
                    f"Line {idx + 1} mismatch: file has {actual!r}, "
                    f"hunk expected {expected!r}"
                )

        # Emit added lines, each with a trailing \n.
        for added in hunk.added_lines:
            new_lines.append(added + "\n")

        # Advance cursor past the removed lines.
        cursor = insert_at + hunk.old_count

    # Copy any remaining lines unchanged.
    new_lines.extend(original_lines[cursor:])
    return "".join(new_lines)

File: web/src/mutation/unifiedDiff.js
// Parsing and application of 0-context unified diffs — a faithful port of
// jadelens/unified_diff.py. Strict, predictable semantics (no fuzz, no line
// drift): all hunks reference the original file's line numbers, removed lines
// must match byte-for-byte, hunks are canonicalised and checked for non-overlap.

export class DiffParseError extends Error {}
export class DiffApplyError extends Error {}

const HUNK_HEADER_RE = /^@@ -(\d+)(?:,(\d+))? \+\d+(?:,(\d+))? @@/;

/** Mirror Python str.splitlines(keepends=True) for `\n`-terminated text. */
function splitlinesKeepends(s) {
  const lines = [];
  let start = 0;
  for (let i = 0; i < s.length; i++) {
    if (s[i] === '\n') {
      lines.push(s.slice(start, i + 1));
      start = i + 1;
    }
  }
  if (start < s.length) lines.push(s.slice(start));
  return lines;
}

function stripTrailingNewline(line) {
  return line.endsWith('\n') ? line.slice(0, -1) : line;
}

export function parseUnifiedDiff(diffText) {
  const lines = diffText.split('\n');
  // A trailing `\n` yields a final empty element from split; drop it.
  if (lines.length && lines[lines.length - 1] === '') lines.pop();

  const hunks = [];
  let i = 0;
  // Skip any preamble (file headers, etc.) until the first hunk header.
  while (i < lines.length && !lines[i].startsWith('@@')) i++;

  while (i < lines.length) {
    const m = HUNK_HEADER_RE.exec(lines[i]);
    if (!m) throw new DiffParseError(`Malformed hunk header: ${JSON.stringify(lines[i])}`);
    const oldStart = parseInt(m[1], 10);
    const oldCount = m[2] !== undefined ? parseInt(m[2], 10) : 1;
    const newCount = m[3] !== undefined ? parseInt(m[3], 10) : 1;
    i++;

    const removed = [];
    const added = [];
    while (i < lines.length && !lines[i].startsWith('@@')) {
      const line = lines[i];
      if (line.startsWith('-')) removed.push(line.slice(1));
      else if (line.startsWith('+')) added.push(line.slice(1));
      else if (line.startsWith(' ')) {
        throw new DiffParseError(
          `Context lines are not supported (0-context format required): ${JSON.stringify(line)}`,
        );
      } else {
        throw new DiffParseError(`Unexpected line in hunk body: ${JSON.stringify(line)}`);
      }
      i++;
    }

    if (removed.length !== oldCount) {
      throw new DiffParseError(
        `Hunk header declared ${oldCount} removed line(s) but hunk body has ${removed.length}`,
      );
    }
    if (added.length !== newCount) {
      throw new DiffParseError(
        `Hunk header declared ${newCount} added line(s) but hunk body has ${added.length}`,
      );
    }

    hunks.push({ oldStart, oldCount, removed, added });
  }

  // Canonicalise: sort by (oldStart, addressing-before-pure-insert), then
  // verify strict ordering / non-overlap (JS Array.sort is stable).
  hunks.sort(
    (a, b) => a.oldStart - b.oldStart || (a.oldCount > 0 ? 0 : 1) - (b.oldCount > 0 ? 0 : 1),
  );
  verifyOrderingAndNonOverlap(hunks);
  return hunks;
}

function verifyOrderingAndNonOverlap(hunks) {
  for (let i = 0; i < hunks.length - 1; i++) {
    const a = hunks[i];
    const b = hunks[i + 1];
    let minBStart;
    if (a.oldCount === 0) {
      minBStart = a.oldStart + 1; // A pure insert; B must start strictly after.
    } else if (b.oldCount === 0) {
      minBStart = a.oldStart + a.oldCount - 1; // insert at the boundary after A is OK.
    } else {
      minBStart = a.oldStart + a.oldCount; // both addressing: B after A's range.
    }
    if (b.oldStart < minBStart) {
      throw new DiffParseError(
        `Hunks out of order or overlapping: previous hunk @@ -${a.oldStart},${a.oldCount} ... @@ ` +
          `is followed by hunk @@ -${b.oldStart},${b.oldCount} ... @@ ` +
          `(B.old_start must be >= ${minBStart})`,
      );
    }
  }
}

export function applyUnifiedDiff(original, diffText) {
  const hunks = parseUnifiedDiff(diffText);
  const originalLines = splitlinesKeepends(original);

  const newLines = [];
  let cursor = 0; // 0-indexed position in originalLines

  for (const hunk of hunks) {
    const insertAt = hunk.oldCount > 0 ? hunk.oldStart - 1 : hunk.oldStart;

    // Copy untouched lines from cursor up to the hunk's position.
    newLines.push(...originalLines.slice(cursor, insertAt));

    // Verify removed lines match file content at the claimed position.
    for (let offset = 0; offset < hunk.removed.length; offset++) {
      const idx = insertAt + offset;
      if (idx >= originalLines.length) {
        throw new DiffApplyError(
          `Hunk references line ${idx + 1} but file only has ${originalLines.length} line(s)`,
        );
      }
      const actual = stripTrailingNewline(originalLines[idx]);
      if (actual !== hunk.removed[offset]) {
        throw new DiffApplyError(
          `Line ${idx + 1} mismatch: file has ${JSON.stringify(actual)}, ` +
            `hunk expected ${JSON.stringify(hunk.removed[offset])}`,
        );
      }
    }

    // Emit added lines, each with a trailing newline.
    for (const addedLine of hunk.added) newLines.push(addedLine + '\n');

    cursor = insertAt + hunk.oldCount;
  }

  newLines.push(...originalLines.slice(cursor));
  return newLines.join('');
}
